api_compatibility_policy.md

```diff
@@ -89,9 +145,16 @@ These changes must be approved by [more than half of the project OWNERS](OWNERS)
 Tekton Pipelines [maintains a list of features that have been deprecated](https://github.com/tektoncd/pipeline/tree/main/docs/deprecations.md)
 which includes the earliest date each feature will be removed.
 
-## Go libraries compatibility policy
+## Go Libraries
+
+The following guidelines describe backwards incompatible changes and how they relate to release versioning for the public Go packages under `github.com/tektoncd/pipeline/pkg`.
+
+Packages that are part of the Go client library for a particular CRD version should adhere to the stability guarantees for that version. This includes the Go structs for the API types in `pkg/apis/pipeline/$apiVersion` and the client libraries generated from those in `pkg/client`. This means:
+
+- Changing or removing a field in a go struct in `pkg/apis/pipeline/v1alpha1` (or another alpha API) is not considered a backwards incompatible change.
+
+- Renaming a field in `pkg/apis/pipeline/v1beta1` (or another beta API) is not considered a backwards incompatible change. However, removing a field is considered a backwards incompatible change and requires a major version bump.
```

Comment:
Similar comment as for alpha. We cannot really define what is and is not backwards compatible.
The [reference](https://go.dev/blog/module-compatibility) you linked below looks like a good one to explain what is and what is not backwards compatible. 

I believe what should change between alpha and beta is the amount of notice we give users for backward incompatible changes. Additionally, we can define changes that require a new major beta version no matter the amount of notice - for instance, we can say that removing a field requires a new major version.

It seems to me that separating renaming from removing is artificial. Renaming is at all effects removal of the old name + the addition of the new name. While the same content is stored, old clients won't find the new field without changes.

I think the way to do renaming is to provide the new name and keep the old one but deprecate it. 
After the agreed deprecation period is met, we remove the old field.

In stable APIs instead, we can never remove a field. 
We could add the new name, which is backwards compatible, but only a new stable version could remove the field

Reply:
I agree that separating removing from renaming is a bit artificial (even though it was partially my idea 😅 ) but I'm not sure how realistic it is to announce to users in advance a change we want to make to the go libs. (maybe it is?)
Do you think it would be OK to just allow backwards incompatible changes in v1beta1 go libs? since the definition of beta is that features won't be removed, but the details may change-- I think if we're not changing the yaml/json definitions it's a reasonable policy.

Reply:
I think it's ok to allow backwards incompatible changes in the beta API, but I think we should provide notice.
Similar to the API policy:
- alpha we can do backwards incompatible without notice 
- beta we can do backwards incompatible changes with notice (i.e. going through deprecation)
- stable we cannot do backwards incompatible changes, they have to go in a new version

Reply:
I tend to agree...what I'm a bit worried about is how well we can actually implement providing notice for the beta  changes - things like adding an extra arg or a return value to a function are not backwards incompatible. Do you see us providing notice for changes like those across all exported functions in pkg/? If so, how long would the notice period be? (Personally, I'd be ok with a `action-required` release note)

Reply:
In terms of how to deprecate, I think action required release notes are good, and comment in the code, so that's shown by IDEs 

In terms of duration, what about three releases? I think our current API policy of 9 months is too much, especially it will be once we have a v1 out, so definitely I would go shorter than that.

Reply:
updated to 3 releases